src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java

```diff
@@ -434,6 +310,210 @@ public void accept(NetworkStateProvider.ConnectionState state) {
+            for (SucceedOnceOperation genCert : certGenOperations) {
+                // cert gen may be expensive, so catch interrupts as early as possible
+                if (Thread.currentThread().isInterrupted()) {
+                    throw new InterruptedException("Interrupted while generating certificates");
+                }
+                genCert.apply();
+            }
+            return null;
+        }
+
+        private void updateCISShadowReportedState(Map<String, Object> reportedState) {
+            UpdateShadowRequest updateShadowRequest = new UpdateShadowRequest();
+            updateShadowRequest.thingName = shadowName.get();
+            updateShadowRequest.state = new ShadowState();
+            updateShadowRequest.state.reported = reportedState == null ? null : new HashMap<>(reportedState);
+            iotShadowClient.PublishUpdateShadow(updateShadowRequest, QualityOfService.AT_LEAST_ONCE)
+                    .exceptionally(e -> {
+                        LOGGER.atDebug().cause(e).log("Unable to update CIS shadow reported state");
+                        return null;
+                    });
+        }
+    }
+
+    class CISShadowTaskExecutor {
+        private Future<?> task;
+
+        @SuppressWarnings("PMD.AvoidCatchingGenericException")
+        void start() {
+            stop();
+            task = executorService.submit(() -> {
+                while (!Thread.currentThread().isInterrupted()) {
+                    ProcessCISShadowTask task = null;
+                    try {
+                        // keep task in queue during processing
+                        task = taskQueue.blockingPeek();
+                        task.call();
+                        taskQueue.removeProcessedTask(task);
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        return;
+                    } catch (Exception e) {
+                        LOGGER.atError().cause(e).kv(KV_TASK, task).log("Unable to process connectivity shadow. "
+                                + "Certificates may not be updated with most current connectivity info. "
+                                + "If components are experiencing TLS handshake errors, restart Greengrass "
+                                + "to refresh certificates");
+                    }
+                }
+            });
+        }
+
+        void stop() {
+            if (task != null) {
+                task.cancel(true);
+            }
+        }
+    }
+
+    /**
+     * A queue for connectivity shadow responses. Old shadows will be rejected from the queue.
+     */
+    class CISShadowTaskQueue  {
+        private final BlockingDeque<ProcessCISShadowTask> tasks = new LinkedBlockingDeque<>();
+        private ProcessCISShadowTask lastProcessedTask;
+        private final Comparator<ProcessCISShadowTask> taskComparator = (task1, task2) -> {
+            if (Objects.equals(task1, task2)) {
```

Comment:
Make sure to handle both null tasks and null versions within those tasks.

Reply:
right now null tasks are filtered out before comparing in `isNewTask`:
```
t -> t == null || taskComparator.compare(task, t) > 0
```
could move it in to the comparator so it's more resilient.
---
null cis versions are already handled through Objects.equals. and shadow versions are just `long`

Reply:
done